enhancements/accelerators/out-of-tree-drivers.md

```diff
@@ -0,0 +1,539 @@
+For prebuilt containers pullable from a vendor's repository we're going to use a ImageContentSourcePolicy, currently only pulling by digest works, we cannot pull by label now. We need to accommodate this in the naming scheme of a DriverContainer.
+If a DriverContainer needs additional RPM packages in a proxy environment we need to update the configuration for dnf and rhsm. 
+ 
+Steps to enable proxy setting for yum, dnf, rhsm in a container 
+
+```
+$ sudo vim /etc/dnf/dnf.conf
+# Add
+proxy=http://proxyserver:port
+
+$ sudo vim /etc/yum.conf
+# Add
+proxy=http://proxyserver:port
+
+$ sudo vi /etc/rhsm/rhsm.conf
+# Configure
+proxy_hostname = proxy.example.com
+proxy_port = 8080
+# user name for authenticating to an http proxy, if needed
+proxy_user =
+
+# password for basic http proxy auth, if needed
+proxy_password =
+```
+
+
+
+### Risks and Mitigations
+
+What are the risks of this proposal and how do we mitigate. Think broadly. For example, consider both security and how this will impact the larger OKD ecosystem.
+
+How will security be reviewed and by whom? How will UX be reviewed and by whom?
+
+Consider including folks that also work outside your immediate sub-project.
+
+## Design Details
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+
+
+### Upgrade / Downgrade Strategy
+
+
+#### Special Resource Driver Updates 
+
+Initially, the NFD operator labels the host with the kernel version (e.g. 4.1.2). The SRO reads this kernel version and creates a DaemonSet with a NodeSelector targeting this kernel version and a corresponding image to pull. With the pci-10de label, the DriverContainer will only land on GPU nodes. 
+
+This way we ensure that an image is pulled and placed only on the node where the kernel matches. It is the responsibility of the DriverContainer image builder to name the image the correct way.
+
+If an administrator updates the CR with a new driver version will taint the node rebuild the DriverContainer and un-taint the node with the new drivers running on a node. 
+
+#### Updates in OpenShift
+
+Updates in OpenShift can happen in two ways: 
+1. Only the payload (operators and needed parts on top of the OS) is updated
+2. The payload and the OS are simultaneously updated. 
+
+The first case is "easy" the new version of the operator will reconcile the expected state and verify that all parts of the special resource stack are working and then "do" nothing. 
```

Comment:
Can you expand on verification?  Is this looking at the status of DriverContainer DaemonSet and/or its Pods to see if they are reporting Ready?

Reply:
SRO can watch for different things via annotations depending on the needs: (1) General availability, resource created (2) All Pods for a DaemonSet are reporting Ready, (3) The log of the Pods in a DaemonSet are checked against a regex see:[https://www.openshift.com/blog/part-2-how-to-enable-hardware-accelerators-on-openshift-sro-building-blocks]( https://www.openshift.com/blog/part-2-how-to-enable-hardware-accelerators-on-openshift-sro-building-blocks)

Reply:
[https://blog.openshift.com/part-1-how-to-enable-hardware-accelerators-on-openshift/](https://blog.openshift.com/part-1-how-to-enable-hardware-accelerators-on-openshift/) This one for completeness @wking